software/fpga/ov3/ovhw/whacker/producer.py

```diff
@@ -182,40 +198,74 @@ def write_hdr(statename, nextname, hdr_offs, val):
                 )
             )
         
+        self.fsm.act("waitfilter",
+            If(filter_done,
+                If(filter_reject,
+                    NextState("IDLE")
+                ).Else(
+                    self.delta_timestamp.set(pkt_timestamp - self.previous_timestamp.v),
+                    self.previous_timestamp.set(pkt_timestamp),
+                    NextState("waitdone")
+                )
+            )
+        )
 
         self.fsm.act("waitdone",
-                If(filter_done,
-                    If(filter_reject,
-                        NextState("IDLE")
-                    ).Else(
-                        clear_acc_flags.eq(1),
-                        NextState("WH0"))
-                ))
-
-        write_hdr("WH0", "WRF0", 0, 0xA0)
+            If(self.delta_timestamp.v[56:64],
+                self.delta_ts_size.set(7),
+                NextState("WRT7")
+            ).Elif(self.delta_timestamp.v[48:56],
+                self.delta_ts_size.set(6),
+                NextState("WRT6")
+            ).Elif(self.delta_timestamp.v[40:48],
+                self.delta_ts_size.set(5),
+                NextState("WRT5")
+            ).Elif(self.delta_timestamp.v[32:40],
+                self.delta_ts_size.set(4),
+                NextState("WRT4")
+            ).Elif(self.delta_timestamp.v[24:32],
+                self.delta_ts_size.set(3),
+                NextState("WRT3")
+            ).Elif(self.delta_timestamp.v[16:24],
+                self.delta_ts_size.set(2),
+                NextState("WRT2")
+            ).Elif(self.delta_timestamp.v[8:16],
+                self.delta_ts_size.set(1),
+                NextState("WRT1")
+            ).Else(
+                self.delta_ts_size.set(0),
+                NextState("WRT0")
+            ),
+            clear_acc_flags.eq(1)
```

Comment:
This will clear `HF0_LAST` even if it was not latched in `self.flags`. This can happen if `CSTREAM_CFG` write to `0` happens while USB packet is being captured. The result is that stuff packet with `HF0_LAST` won't ever happen.

Reply:
Fixed by clearing flags only if the relevant flags were captured at start of packet. This guarantees that `HF0_LAST` will arrive after stopping capture.